### PR TITLE
BUG-24971 copying blocks also considers ndim

### LINKED ISSUE #24971: Categorical.replace returns unexpected dimensions for length 1 Series
The shapes and types of the first two are unexpected. Should they be? Just upgraded to pandas 0.24.0.

```python
pd.Series(pd.Categorical('A', categories=['A', 'B'])).replace({'A': 1, 'B': 2})
# Out[30]:
# 0    [1]
# dtype: object

pd.Series(pd.Categorical(('A', ), categories=['A', 'B'])).replace({'A': 1, 'B': 2})
# Out[31]:
# 0    [1]
# dtype: object

pd.Series(pd.Categorical(('A', 'B'), categories=['A', 'B'])).replace({'A': 1, 'B': 2})
# Out[32]:
# 0    1
# 1    2
# dtype: int64
```

I expect the first to be an int64 dtype with a shape of (1,). They return a Series with shape (1, 1) and dtype object.

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-116-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0
pytest: 4.1.1
pip: 18.1
setuptools: 40.2.0
Cython: None
numpy: 1.15.4
scipy: 1.2.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2018.7
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.2.15
pymysql: None
psycopg2: 2.7.6.1 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #24971 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

The following gives a series containing [1] instead of 1
```
>>> pd.Series(pd.Categorical('A', categories=['A', 'B'])).replace({'A': 1, 'B': 2})
0    [1]
dtype: object
```
This bug occurs because in the process of copying the original categorical block (which is needed as the operation is not inplace), the constructor class for the new object defaults to `ObjectBlock`, whose constructor has a default `ndim` of 2. This PR alters the block `copy` function to specify that the newly constructed block should have the same `ndim` as the block being copied.

### CODE REVIEW

File: pandas/tests/series/test_constructors.py [hunk 1]
```diff
@@ -532,6 +532,11 @@ def test_constructor_copy(self):
             assert x[0] == 2.
             assert y[0] == 1.
 
+    def test_copy_categorical_ndim(self):
+        # GH 24971
+        s = Series(Categorical(['A'], categories=['A']))
+        assert not is_list_like(s.replace({'A': 1})[0])
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: use assert_series_equal

File: doc/source/whatsnew/v0.24.2.rst [hunk 2]
```diff
@@ -66,7 +66,7 @@ Bug Fixes
 
 **Categorical**
 
--
+- Bug where a copy of a categorical could have the wrong dimensions (:issue:`24971`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.2.rst]: can you reword to make this more clear. mentione this is on a Series of categorical dtype. The dimenension issue is an internal one. You want to emphasize the user visible effects.

Comment (jreback) [reply doc/source/whatsnew/v0.24.2.rst]: this is visible when you call ``.replace()``

File: pandas/tests/series/test_constructors.py [hunk 3]
```diff
@@ -532,6 +532,11 @@ def test_constructor_copy(self):
             assert x[0] == 2.
             assert y[0] == 1.
 
+    def test_copy_categorical_ndim(self):
+        # GH 24971
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: can you put all 3 tests from the OP here.

File: pandas/tests/series/test_constructors.py [hunk 4]
```diff
@@ -532,6 +532,11 @@ def test_constructor_copy(self):
...
+        s = Series(Categorical(['A'], categories=['A']))
+        assert not is_list_like(s.replace({'A': 1})[0])
+
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: this should go with the replace tests

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by JustinZhengBC** [comment]
@jreback I've made the requested changes

File: pandas/tests/series/test_replace.py [hunk 5]
```diff
@@ -280,3 +280,17 @@ def test_replace_mixed_types_with_string(self):
...
+        (pd.Categorical(('A', 'B'), categories=['A', 'B']), [1, 2]),
+    ])
+    def test_copy_categorical_ndim(self, categorical, numeric):
+        # GH 24971
+        s = pd.Series(categorical)
+        result = s.replace({'A': 1, 'B': 2})
+        expected = pd.Series(numeric)
+        print(result.dtype)
+        print(expected.dtype)
+        tm.assert_series_equal(expected, result, check_dtype=False)
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/tests/series/test_replace.py]: Looks like L292 has the reference now (https://github.com/pandas-dev/pandas/issues/23305)

**Comment by jreback** [comment]
can you merge master

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @JustinZhengBC nice patch, keep em coming!

**Pull request merged**